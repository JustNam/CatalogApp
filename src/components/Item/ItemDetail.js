import React, { Component } from 'react';
import { ButtonToolbar, Button } from 'react-bootstrap';
import EditItemModal from '../Modals/EditItemModal';
import SuccessModal from '../Modals/SuccessModal';
import ConfirmModal from '../Modals/ConfirmModal';
import NavigationBar from '../HomePage/NavigationBar';
import { historyWithRefresh } from '../../history';

class ItemDetail extends Component {
  constructor(props) {
    super(props);

    const { categoryId, itemId } = this.props.match.params;
    this.state = {
      showInfoModal: false,
      showSuccessModal: false,
      showConfirmModal: false,
      categoryId,
      itemId,
      backToHome: null,
      itemDetail: {},
    };
  }

  async componentDidMount() {
    // Prevent refresh from crashing
    const { categoryId, itemId } = this.props.match.params;
    await this.props.getItemInCategory(categoryId, itemId);
    const { item } = this.props;
    this.setState({ itemDetail: item.data[0] });
  }

  deleteItem = (e) => {
    e.preventDefault();
    const { categoryId, itemId } = this.state;
    this.props.deleteItemInCategory(categoryId, itemId).then(() => {
      this.handleConfirmClose();
      this.handleSuccessShow();
    });
  }

  handleInfoShow = () => {
    this.setState({ showInfoModal: true });
    this.setState({ backToHome: false });
  };

  handleInfoClose = () => {
    this.setState({ showInfoModal: false });
  };

  handleSuccessShow = () => {
    this.setState({ showSuccessModal: true });
  };

  handleSuccessClose = () => {
    const { categoryId, itemId } = this.state;
    this.setState({ showSuccessModal: false });
    if (this.state.backToHome) {
      historyWithRefresh.push('/categories');
    } else {
      historyWithRefresh.push(`/categories/${categoryId}/items/${itemId}`);
    }
  };

  handleConfirmShow = () => {
    this.setState({ showConfirmModal: true });
    this.setState({ backToHome: true });
  };

  handleConfirmClose = () => {
    this.setState({ showConfirmModal: false });
  };

  render() {
    const { itemDetail } = this.state;
    const userId = localStorage.getItem('userId');
    if (itemDetail.user) {
      const { showInfoModal, showSuccessModal, showConfirmModal, categoryId } = this.state;
      return (
        <div>
          <NavigationBar />
          <div className="container page">
            <div className="article-preview">
              <div className="article-meta">
                <div className="info">
                  <div className="author">
                    {itemDetail.user.username}
                  </div>
                  <span className="date">{`updated ${itemDetail.created_on.replace(',', ' at')}, created ${itemDetail.updated_on.replace(',', ' at')}`}</span>
                </div>
              </div>
              <div className="preview-link">
                <h1>{itemDetail.title}</h1>
                <p>{itemDetail.description}</p>
                {!itemDetail.description && <p><i>No description</i></p>}
                {(parseInt(userId) === parseInt(itemDetail.user.id))
                  && (
                  <span>
                    <ButtonToolbar className="modal-footer">
                      <Button
                        id="editButton"
                        variant="outline-success"
                        size="sm"
                        onClick={this.handleInfoShow}
                      >
                        Edit
                      </Button>
                      <Button
                        variant="outline-success"
                        size="sm"
                        onClick={this.handleConfirmShow}
                      >
                      Delete
                      </Button>
                    </ButtonToolbar>
                  </span>
                  )
                }
              </div>
            </div>
          </div>

          <EditItemModal
            currentCategoryId={categoryId}
            item={itemDetail}
            show={showInfoModal}
            handleClose={this.handleInfoClose}
            showSuccessModal={this.handleSuccessShow}
          />
          <SuccessModal show={showSuccessModal} handleClose={this.handleSuccessClose} />
          <ConfirmModal
            show={showConfirmModal}
            handleConfirm={this.deleteItem}
            handleClose={this.handleConfirmClose}
          />
        </div>
      );
    }

    return <div><NavigationBar /></div>;
  }
}
export default ItemDetail;