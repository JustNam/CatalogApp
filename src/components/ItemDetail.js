import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

class ItemDetail extends Component {
  componentDidMount() {
    // eslint-disable-next-line react/destructuring-assignment
    const { categoryId, itemId } = this.props.match.params;
    console.log(categoryId);
    console.log(itemId);
    // eslint-disable-next-line react/destructuring-assignment
    this.props.getItemInCategory(categoryId, itemId);
  }

  render() {
    const { item } = this.props;
    const itemDetail = item.data[0];
    if (itemDetail) {
      return (
        <div>
          <div className="article-preview">
            <div className="article-meta">
              <div className="info">
                <div className="author">
                  {itemDetail.user.username}
                </div>
                <span className="date">{`updated ${itemDetail.created_on.replace(',', ' at')}, created ${itemDetail.updated_on.replace(',', ' at')}`}</span>
              </div>
            </div>
            <div className="preview-link" href="#article/books-4kbydz">
              <h1>{itemDetail.title}</h1>
              <p>{itemDetail.description}</p>
              {!itemDetail.description && <p><i>No description</i></p>}
            </div>
          </div>
        </div>
      );
    }
    return <div />;
  }
}

const mapStateToProps = state => ({});
const mapDispatchToProps = dispatch => ({});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ItemDetail));
