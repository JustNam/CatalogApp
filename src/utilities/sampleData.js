export const errorItem = {
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 7,
      title: 'Ru4)',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const itemWithEmptyCategory = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 1,
  data: [],
};

export const noDescriptionItem = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 1,
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 7,
      title: 'Ru4',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const item = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 3,
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      description: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum',
      id: 7,
      title: 'Ru4',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 8,
      title: 'Ru3',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      description: 'Lorem Ipsum is simply',
      id: 9,
      title: 'Ru2',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const category = {
  data: [
    {
      created_on: '06/26/2019, 08:09:32',
      id: 1,
      name: 'Swimming',
      updated_on: '06/26/2019, 08:09:32',
    },
    {
      created_on: '06/26/2019, 08:09:32',
      id: 2,
      name: 'Running',
      updated_on: '06/26/2019, 08:09:32',
    },
  ],
};
