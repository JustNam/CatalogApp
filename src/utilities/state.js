export const initalCategoryState = {
  data: [],
};

export const initalItemState = {
  data: [],
  categoryId: 0,
  currentPage: 0,
  lastPage: 0,
};

export const initalUserState = {
  loggedIn: localStorage.getItem('accessToken') || false,
};