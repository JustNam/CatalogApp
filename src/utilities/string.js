export const convertCamelCaseToSnakeCase = string => (
  string.replace(/[A-Z]/g, match => `_${match.toLowerCase()}`)
);