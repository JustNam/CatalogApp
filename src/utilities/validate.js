export const validateItemTitle = (title) => {
  const { value } = title;
  const errors = [];
  if (!value.match(/^[a-zA-Z0-9\s]+$/)) {
    errors.push('Given title contains invalid character');
  }
  return {
    value,
    isValid: (errors.length === 0),
    errorMessages: errors,
  };
};

export const validateUsername = (username) => {
  const { value } = username;
  const errors = [];
  if (!value.match(/^[a-zA-Z0-9]+$/)) {
    errors.push('Given username contains invalid character');
  }
  if (!value.length > 30) {
    errors.push('Given username contains more than 30 characters');
  }
  return {
    value,
    isValid: (errors.length === 0),
    errorMessages: errors,
  };
};