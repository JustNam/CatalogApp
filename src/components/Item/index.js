import ItemDetail from 'components/Item/ItemDetail';

export default ItemDetail;