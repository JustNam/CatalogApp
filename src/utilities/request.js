import { convertCamelCaseToSnakeCase } from './string';
import envConfig from '../configurations';

export const convertToRequest = (dict) => {
  const newDict = {};
  Object.keys(dict).forEach((key) => {
    const newKey = convertCamelCaseToSnakeCase(key);
    newDict[newKey] = dict[key];
  });

  return JSON.stringify(newDict);
};

export const callAPI = async (enpoint, method, body, customizedHeaders = {}) => {
  const defaultHeaders = {
    'Content-Type': 'application/json',
  };
  const headers = {
    ...defaultHeaders,
    ...customizedHeaders,
  };

  let tempBody;
  if (body) {
    if (headers['Content-Type'] === 'application/json') {
      // Eliminate fields which is null
      const processedBody = body;
      Object.keys(processedBody).forEach((key) => {
        if (!processedBody[key]) {
          processedBody[key] = '';
        }
      });
      tempBody = convertToRequest(processedBody);
    } else {
      delete headers['Content-Type'];
      tempBody = body;
    }
  } else {
    delete headers['Content-Type'];
  }
  if (localStorage.getItem('accessToken')) {
    headers.authorization = `Bearer ${localStorage.getItem('accessToken')}`;
  }

  const config = {
    method,
    headers,
  };

  // Only allow the other endpoints to assign body
  if (method !== 'GET') {
    config.body = tempBody;
  }

  function handleErrors(response) {
    if (!response.ok) {
      throw response;
    }
    return response;
  }
  const result = await fetch(envConfig.domain.concat(enpoint), config)
    .then(handleErrors)
    .then(response => response.json());
  return result;
};