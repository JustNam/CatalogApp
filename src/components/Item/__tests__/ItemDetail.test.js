import React from 'react';
import { shallow } from 'enzyme';
import { ItemDetail } from 'components/Item/ItemDetail';
import { item, itemWithEmptyCategory, noDescriptionItem } from 'utilities/sampleData';


describe('ItemDetail with available item', () => {
  let wrapper;
  let props;
  let instance;
  const setup = () => {
    props = {
      match: {
        params: {
          categoryId: 1,
          itemId: 7,
        },
      },
      item: {
        data: [
          {
            id: 1,
          },
        ],
      },
      itemDetail: {
        created_on: 1,
      },
      showInfoModal: false,
      showSuccessModal: false,
      showConfirmModal: false,
      deleteItemInCategory: jest.fn(() => Promise.resolve()),
      getItemsInCategory: jest.fn(() => Promise.resolve()),
    };
  };
  beforeEach(() => {
    localStorage.setItem('userId', 1);
    setup();
    wrapper = shallow(<ItemDetail {...props} />);
    instance = wrapper.instance();
  });
  it('It should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('It should show a CreateItemModal', () => {
    wrapper.setState({
      ...wrapper.state(),
      itemDetail: {
        user: {
          id: 1,
        },
        created_on: 'test',
        updated_on: 'test',
      },
    });
    localStorage.setItem('userId', 1);
    wrapper.find('Button[id="editButton"]').simulate('click');
    expect(wrapper.state().showInfoModal).toEqual(true);
    localStorage.clear();
  });
  it('It should close CreateItemModal when handleInfoClose is called', () => {
    instance.handleInfoClose();
    expect(wrapper.state().showInfoModal).toEqual(false);
  });
  it('It should open SuccessModal when handleSuccessShow is called', () => {
    instance.handleSuccessShow();
    expect(wrapper.state().showSuccessModal).toEqual(true);
  });
  it('It should close SuccessModal when handleSuccessClose is called', () => {
    instance.handleSuccessClose();
    expect(wrapper.state().showInfoModal).toEqual(false);
  });
  it('It should redirect to homepage when handleSuccessClose is called from delete function', () => {
    wrapper.setState({
      backToHome: true,
    });
    instance.handleSuccessClose();
    expect(wrapper.state().showInfoModal).toEqual(false);
  });
  it('It should open ConfirmModal when handleConfirmShow is called', () => {
    instance.handleConfirmShow();
    expect(wrapper.state().showConfirmModal).toEqual(true);
  });
  it('It should close ConfirmModal when handleConfirmClose is called', () => {
    instance.handleConfirmClose();
    expect(wrapper.state().showConfirmModal).toEqual(false);
  });
  it('It should show SuccessModal when an item is deleted', () => {
    instance.deleteItem({
      preventDefault: () => {},
    });
    expect(wrapper.state().showConfirmModal).toEqual(false);
  });
});

describe('ItemDetail with item which does not exist', () => {
  let wrapper;
  let props;
  let instance;
  const setup = () => {
    props = {
      match: {
        params: {
          categoryId: 1,
          itemId: 7,
        },
      },
      item: itemWithEmptyCategory,
      showInfoModal: false,
      showSuccessModal: false,
      showConfirmModal: false,
      deleteItemInCategory: jest.fn(() => Promise.resolve()),
      getItemsInCategory: jest.fn(() => Promise.resolve()),
    };
  };
  beforeEach(() => {
    localStorage.setItem('userId', 1);
    setup();
    wrapper = shallow(<ItemDetail {...props} />);
    instance = wrapper.instance();
  });
  it('It should show SuccessModal when an item is deleted', async () => {
    await instance.deleteItem({
      preventDefault: () => {},
    });
    expect(wrapper.state().showSuccessModal).toEqual(true);
  });
});

describe('ItemDetail with item which does not have description', () => {
  let wrapper;
  let props;
  const setup = () => {
    props = {
      match: {
        params: {
          categoryId: 1,
          itemId: 7,
        },
      },
      item: noDescriptionItem,
      showInfoModal: false,
      showSuccessModal: false,
      showConfirmModal: false,
      deleteItemInCategory: jest.fn(() => Promise.resolve()),
      getItemsInCategory: jest.fn(() => Promise.resolve()),
      itemDetail: {
        user: {
          id: 1,
        },
        created_on: 'test',
        updated_on: 'test',
      },
    };
  };
  beforeEach(() => {
    localStorage.setItem('userId', 1);
    setup();
    wrapper = shallow(<ItemDetail {...props} />);
  });
  it('It should notify that the item does not have description', async () => {
    wrapper.setState({
      ...wrapper.state(),
      itemDetail: {
        user: {
          id: 1,
        },
        created_on: 'test',
        updated_on: 'test',
      },
    });
    expect(wrapper.find('i').text()).toContain('No description');
  });
});