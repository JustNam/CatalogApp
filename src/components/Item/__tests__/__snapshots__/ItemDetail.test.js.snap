// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetail with available item It should render correctly 1`] = `ShallowWrapper {}`;
